test/UnichainProtocolFees.fork.t.sol

```diff
@@ -0,0 +1,217 @@
+      vm.deal(address(this), feeAmount);
+      (bool success,) = address(tokenJar).call{value: feeAmount}("");
+      assertTrue(success, "Failed to send ETH to TokenJar");
+
+      // Deal UNI and release
+      deal(RESOURCE, callers[i], THRESHOLD);
+
+      uint256 _nonce = releaser.nonce();
+      Currency[] memory currencies = new Currency[](1);
+      currencies[0] = Currency.wrap(address(0)); // ETH
+
+      vm.startPrank(callers[i]);
+      IERC20(RESOURCE).approve(address(releaser), THRESHOLD);
+
+      uint256 recipientBalanceBefore = callers[i].balance;
+      releaser.release(_nonce, currencies, callers[i]);
+
+      // Verify release
+      assertEq(callers[i].balance - recipientBalanceBefore, feeAmount, "Incorrect ETH released");
+      assertEq(address(tokenJar).balance, 0, "TokenJar not emptied");
+      vm.stopPrank();
+    }
+  }
+
+  function test_ownershipTransfer() public {
+    // Test that ownership can be transferred by current owner
+    address newOwner = address(0x9999);
+
+    // Transfer TokenJar ownership
+    vm.prank(owner);
+    IOwned(address(tokenJar)).transferOwnership(newOwner);
+    assertEq(IOwned(address(tokenJar)).owner(), newOwner, "TokenJar ownership not transferred");
+
+    // Transfer Releaser ownership
+    vm.prank(owner);
+    IOwned(address(releaser)).transferOwnership(newOwner);
+    assertEq(IOwned(address(releaser)).owner(), newOwner, "Releaser ownership not transferred");
+
+    // Transfer threshold setter
+    vm.prank(newOwner);
+    releaser.setThresholdSetter(newOwner);
+    assertEq(releaser.thresholdSetter(), newOwner, "Threshold setter not transferred");
+  }
+
+  function test_thresholdUpdate() public {
+    // Test that threshold can be updated by thresholdSetter
+    uint256 newThreshold = 20_000e18;
+
+    vm.prank(owner);
+    releaser.setThreshold(newThreshold);
+    assertEq(releaser.threshold(), newThreshold, "Threshold not updated");
+  }
+
+  function test_releaserUpdate() public {
+    // Test that releaser can be updated on TokenJar
+    address newReleaser = address(0x8888);
+
+    vm.prank(owner);
+    tokenJar.setReleaser(newReleaser);
+    assertEq(tokenJar.releaser(), newReleaser, "Releaser not updated on TokenJar");
+  }
+
+  function test_invalidRelease_insufficientUNI() public {
+    // Test that release fails without sufficient UNI
+    address caller = address(0x7777);
+
+    // Send ETH to TokenJar
+    vm.deal(address(this), 1 ether);
+    (bool success,) = address(tokenJar).call{value: 1 ether}("");
+    assertTrue(success);
+
+    // Give caller less than threshold UNI
+    deal(RESOURCE, caller, THRESHOLD - 1);
+
+    uint256 _nonce = releaser.nonce();
+    Currency[] memory currencies = new Currency[](1);
+    currencies[0] = Currency.wrap(address(0));
+
+    vm.startPrank(caller);
+    IERC20(RESOURCE).approve(address(releaser), THRESHOLD - 1);
+
+    // Should revert due to insufficient UNI
+    vm.expectRevert();
```

Comment:
i think we should max approve here and use `vm.expectRevert(address(RESOURCE))` to verify that the revert is from an insufficient balance and not insufficient approval

Reply:
went ahead and fixed this